feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,106 @@
+
+### Improving workflow user awareness: A workflow file template
+
+In order to increase workflow usage, we also need to make users aware of them. To this end, it is worth considering adding a template workflow file to the default Renku project structure. For example, this could be a workflow file with a few example workflow steps written out but commented out. The user can uncomment it and replace the placeholders with their command, inputs, output, etc. There could also be a comment at the top of the file describing how to run the workflow.
+
+This would put workflows in front of more users, increasing awareness and hopefully also adoption. 
+
+### UX Note: Re-runnability & Pre-existing Outputs
+
+A common frustration users have with current Renku workflows is that the `workflow` & `run` (mostly `run`) commands errors when you run a workflow command that writes an output file that already exists.
+
+Users do this frequently during development: run script `a`, generate file `b`. Change script `a`, rerun, generate a new b. Great, now you're happy with script `a`, so `renku run script_a` one last time to record it. But now Renku complains that no outputs were generated (because file `b` already existed, so nothing was detected)! Users *can* avoid this by telling Renku explicitly that the output is file `b`, but that's more typing. More commonly, users end up deleting output files before every `renku run`, which they- very reasonably- don't like doing.
+
+The file-based workflow feature should solve this. The workflow should read the specified output in the workflow file, rather than detecting changed files. This removes the need for the user to delete the file so that Renku can detect its creation.
+
+
+## 🐰 Rabbit Holes
+
+### Venturing into the Metadata Realm
+
+This pitch will require a substantial effort to make the metadata all play nice. Without diving into implementation details, here are a few high-level requirments that the metadata must meet in order to provide a good user experience:
+
+* The workflows defined in files should be fully represented in the Renku KG metadata.
+* However, the workflow file itself should be the single source of truth for the workflow definition (not the KG metadata).
+* File based workflows can only be edited via the file. These workflows cannot be edited via the CLI or (in the future) somewhere in the RenkuLab UI.
+* In terms of versioning a file-based workflow, the metadata should contain metadata about the file-based workflow at each execution. There is no need to maintain a history of the file-based workflow at commits where the workflow was modified but not executed.
```

Comment:
This first sentence is confusing. You mean to say that the renku KG should capture the file-based WF at execution _only_? I.e. it should not automagically interpret changes to the file via a hook mechanism?

Reply:
This one comes from me. We probably don't want to
- have a githook to pick up changes and create new plans
- require the user to run a command to update based on an edit 
- walk git history to create plans on the fly for all commits to the file on graph generation (the thought alone of going back to that gives me nightmares).

So the compromise with good UX and technical feasibility is to create a plan object in the repo metadata on execution (if one doesn't exist already) and to also create one on the fly for any modifications that weren't executed, but only for the latest version (no walking of git history or anything). This latest version would then be mutable whereas the rest isn't. 
This is so users don't have to run a special command to update the KG metadata but to also see their most recent edits in the UI (not having them show up would be confusing).

We could leave it open for a hook based approach, but then I don't like adding crucial functionality to the hooks, since they might not be set up and a user might skip them for other reasons (like `pre-commit` failing for a non-issue).

We could still have a command to update metadata from file without execution like `renku workflow update` or something, but that's more of a detail

Reply:
yep, that makes sense and it's how I understood it but it took me a while to parse the "meta-meta-metadata" piece. 😆

Reply:
Yes